### PR TITLE
Update type definitions

### PR DESCRIPTION
**Pull request by victorhmp**
#### What problem is this solving?

Currently, the defined types in this app are not exported correctly by our React Builder, since it can't pick them up. This means that anyone who is using this app doesn't get any types from it, and has to type the functions themselves.

This PR aims to fix this issue, and also add a description to the most used entry point for this app, the `useProduct` hook.

#### How to test it?

This is more of an internal change, so nothing visible should be different. This app is linked in this [Workspace](https://victormiranda2--storecomponents.myvtex.com/traveler-backpack/p), along with `vtex.product-gifts`, which is using the newly exported types 😄 .

#### Screenshots or example usage:

Screen Shot 2020-08-31 at 11 44 56

Screen Shot 2020-08-31 at 11 51 42

### CODE REVIEW

**Comment by victorhmp** [comment]
> Can you create a `useProductDispatch()`?

Sure! I've just added it :)

I'll update the documentation PR to reflect this.

File: react/ProductContext.ts [hunk 1]
```diff
@@ -1,3 +1,9 @@
 import { createContext } from 'react'
 
-export default createContext({})
+import { ProductContextState } from './ProductContextProvider'
+
+const productContext = createContext<Partial<ProductContextState> | undefined>(
```
Comment (Klynger) [inline react/ProductContext.ts:5]: Maybe capital like [the `SliderContext`](https://github.com/vtex-apps/slider-layout/blob/master/react/components/SliderContext.tsx#L112)
```suggestion
const ProductContext = createContext<Partial<ProductContextState> | undefined>(
```

File: react/ProductContextProvider.tsx [hunk 2]
```diff
@@ -4,6 +4,76 @@ import ProductContext from './ProductContext'
...
+
+export interface ProductContextState {
+  selectedItem?: Item | null
+  product: MaybeProduct
+  selectedQuantity: number
+  skuSelector: {
+    selectedImageVariationSKU: string | null
+    isVisible: boolean
+    areAllVariationsSelected: boolean
+  }
```
Comment (Klynger) [inline react/ProductContextProvider.tsx:39]: Maybe these attributes of `skuSelector` should be partial too, just for safity

File: react/ProductContextProvider.tsx [hunk 3]
```diff
@@ -4,6 +4,76 @@ import ProductContext from './ProductContext'
...
+  buyButton: BuyButtonContextState
+  assemblyOptions: {
+    items: Record<GroupId, AssemblyOptionItem[]>
+    inputValues: Record<GroupId, InputValues>
+    areGroupsValid: Record<GroupId, boolean>
+  }
```
Comment (Klynger) [inline react/ProductContextProvider.tsx:45]: I don't know about these ones, but could you confirm that these are never `undefined` or `null`?

File: react/ProductContext.ts (hunk 1 above)
Comment (victorhmp) [reply react/ProductContext.ts:5]: Niice

File: react/ProductContextProvider.tsx (hunk 3 above)
Comment (victorhmp) [reply react/ProductContextProvider.tsx:45]: Yep! They could be empty objects though. But I guess the Record type would already cover that, right?

File: react/ProductContextProvider.tsx (hunk 2 above)
Comment (victorhmp) [reply react/ProductContextProvider.tsx:39]: Ok! No problem 😄

**Review by kaisermann — APPROVED** [review]

File: react/ProductContextProvider.tsx (hunk 3 above)
Comment (Klynger) [reply react/ProductContextProvider.tsx:47]: The `Record` can only be an empty object, but it doesn't cover if `assemblyOptions.items === undefined` for instance

Comment (Klynger) [reply react/ProductContextProvider.tsx:47]: Is that ok?

Comment (victorhmp) [reply react/ProductContextProvider.tsx:47]: Yeep, no problem

**Review by Klynger — APPROVED** [review]

**Pull request merged**